design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,724 @@
+user (system:authenticated), bound to particular users, or a mixture of both.
+
+When the policy controller evaluates whether a particular policy should be
+checked against a CertificateRequest, it will perform a SubjectAccessReview
+against the user present on the CertificateRequest. Since the
+CertificateRequestPolicy can be bound by both namespace and cluster scope, the
+SubjectAccessReview will be requested for both the namespace the
+CertificateRequest resides in, as well as at the cluster scope. The namespace
+scope will be checked first in an effort to improve performance (a large number
+of CertificateRequestPolicy resources are likely scoped by namespace). The
+SubjectAccessReview will be requested for every single CertificateRequestPolicy
+that exists, until and if a CertificateRequestPolicy which is bound approves the
+request. If no CertificateRequestPolicy approves the request, the
+CertificateRequest is marked as denied.
+
+The SubjectAccessReview will request for the attributes:
+
+```yaml
+Group: policy.cert-manager.io
+Verb: use
+Resource: CertificateRequestPolicy
+Name: <Name of the CertificateRequestPolicy being checked>
+Namespace: <Namespace of the CertificateRequest, empty if checking cluster scope>
+```
+
+---
+
+## External Policy Server
+
+To allow for third parties to implement extensions to policy that the core
+functionality does not provide, the CertificateRequestPolicy resource can list
+ExternalPolicyServers. An ExternalPolicyServer resource represents a single
+policy engine server that can receive policy checks, evaluate the request, and
+then respond as to whether the request is approved, not-approved, or in an error
+state.
+
+
+For each ExternalPolicyServer listed in the CertificateRequestPolicy, the policy
+controller will request from each server, where any not-approved request will
+make the entire CertificateRequestPolicy as not-approved. All
+ExternalPolicyServer may be called in turn, or perhaps concurrently to improve
+performance. Connections will be made to the ExternalPolicyServers using gRPC.
+This connection is secured using TLS, where the server's CA certificate is
+discovered by the ExternalPolicyServer resource.
+
+
+If the call to the ExternalPolicyServer fails to return a result (network error,
+incorrect CA, server not ready), then the CertificateRequestPolicy will return a
+pending state. The policy controller will retry this CertificateRequestPolicy
+for the same request at a later time if no other CertificateRequestPolicys
+exist, they deny the request, or also return a pending state (i.e. no other
+CertificateRequestPolicys approve the request).
+
+We may want to introduce a 4th response type "defer" to capture the idea that
+the ExternalPolicyServer does not handle the request given and cant compute it.
+This is functionality equivalent to "approve", however bubbles up the intent
+better to statuses.
+
+### API
+
```

Comment:
I think it makes sense to introduce the API with a few examples. The Go definition is stricter and is semantically better, but the reader won't immediately "have a sense" of what the API is like with just the Go definition

I propose to add one or two examples **before** we give the formal Go definition.

@JoshVanL what do you think?

~~~markdown

The External Policy Server API may be used for implementing a custom policy server.
Using the External Policy Server API for that purpose looks like:


```yaml
kind: Deployment
metadata:
  name: custom-policy-engine
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: custom-policy-engine
  template:
    metadata:
      labels:
        app.kubernetes.io/name: custom-policy-engine
    spec:
      containers:
      - name: custom-policy-engine
        image: custom-policy-engine:0.1.0
        ports:
        - containerPort: 8080
---
kind: Service
spec:
  type: ClusterIP
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
  selector:
    app.kubernetes.io/name: custom-policy-engine
---
kind: ExternalPolicyServer
metadata:
  name: opa-shim
spec:
  endpoint: https://custom-policy-engine:8080
  caBundle:                  # optional
    data: |                  # optional
      base64-encoded-pem-file
    configMapRef:            # optional
      name: opa-shim-config
    secretRef:               # optional
      name: opa-shim-tls
status:
  conditions:
  - type: Ready
    status: "True"
    lastTransitionTime: "2021-02-18T06:53:01Z"
    reason: "OpaShimReady"
    message: "opa-shim is ready"
    observedGeneration: 42
```

We aim at making it easy to plug an existing policy engine, such as OPA, using a shim
server that would do the translation work between the External Policy Server API and
the OPA API. The shim, probably a Go binary running as a Deployment, would expose
the `https://opa-shim:8080` endpoint (for example).

```yaml
kind: ExternalPolicyServer
metadata:
  name: opa-shim
spec:
  endpoint: https://opa-shim:8080
  caBundle:          
    data: |              
      base64-encoded-pem-file
    configMapRef:
      name: opa-shim-config
    secretRef:       
      name: opa-shim-tls
status:
  conditions:
  - type: Ready
    status: "True"
    lastTransitionTime: "2021-02-18T06:53:01Z"
    reason: "OpaShimReady"
    message: "opa-shim is ready"
    observedGeneration: 42
```

~~~

Reply:
@JoshVanL

Reply:
Thanks @maelvls, I will add this as an appendix as an example ExternalPolicyServer implementation.